pkg/auth/user/util.go

```diff
@@ -25,3 +32,52 @@ func ConvertAttributes(attrMap map[string][]string) []*v1.UserAttribute {
 	})
 	return result
 }
+
+type loggableAuthProvider struct {
+	Id   string
+	Name string
+	Type string
+}
+
+func protoToJSON(message proto.Message) string {
+	result, err := jsonutil.ProtoToJSON(message, jsonutil.OptCompact, jsonutil.OptUnEscape)
+	if err != nil {
+		utilLog.Error("Failed to convert proto to JSON: ", err)
+		return ""
+	}
+	return result
+}
+
+// LogSuccessfulUserLogin logs user attributes in the specified logger instance.
+func LogSuccessfulUserLogin(log *logging.Logger, user *v1.AuthStatus) {
+	log.Warnw("User successfully logged in with user attributes", extractUserLogFields(user)...)
+}
+
+// The reason this function returns []interface{} instead of []zap.Field
+// is because log.Warnw accepts ...interface{} and []zap.Field does not convert automatically
+// to []interface{}.
+func extractUserLogFields(user *v1.AuthStatus) []interface{} {
+	serviceIdStr := ""
+	permissionsStr := ""
+	if user.GetServiceId() != nil {
+		serviceIdStr = protoToJSON(user.GetServiceId())
```

Comment:
Maybe check for `nil` in `protoToJSON` and return empty string in that case? Then you can save a few lines here and the function is better protected against nil pointer dereferences.

Reply:
I already tried that and didn't succeed - apparently in the Golang interface "hides" `nil` and check like `proto == nil` wouldn't work. We would need to use reflection there, and I think it's better to avoid that: https://stackoverflow.com/questions/13476349/check-for-nil-and-nil-interface-in-go

Reply:
Hm that is unfortunate.